modules/svg/data.js

```diff
@@ -345,15 +363,26 @@ export function svgData(projection, context, dispatch) {
             case '.json':
                 gj = JSON.parse(data);
                 break;
+            case '.shp':
+                shapeToGeoJson(data,_callback);
+                break;
         }
 
-        if (!_isEmpty(gj)) {
-            _geojson = ensureIDs(gj);
-            _src = extension + ' data file';
-            this.fitZoom();
+        function _callback(res){
+            if (res){
+                gj = res;
+            }
+
+            if (!_isEmpty(gj)) {
+                _geojson = ensureIDs(gj);
+                _src = extension + ' data file';
+                drawData.fitZoom();
+            }
+    
+            dispatch.call('change');
         }
 
-        dispatch.call('change');
+        _callback();
```

Comment:
Looks like `_callback()` is always synchronously called here, even in the async case where `shapeToGeoJson` calls it later with actual data. 

Can you rename `_callback` to `done`, and then move this lone call after the `gj=JSON.parse(data)`.  I just think it would be clearer what the code does if the "finalizer" is called only once.

Reply:
changed